### PR TITLE
Use `{cli}` progress bar

### PR DESCRIPTION
**Pull request by IndrajeetPatil**
TODO:

- [x] check if lifecycle is still experimental (https://github.com/r-lib/cli/issues/709)
- [x] ~~figure out how to test this feature~~

### CODE REVIEW

File: R/lint.R [hunk 1]
```diff
@@ -182,20 +182,11 @@ lint_dir <- function(path = ".", ...,
...
-  }
-
   lints <- flatten_lints(lapply(
-    files,
-    function(file) {
-      maybe_report_progress(pb)
-      lint(file, ..., parse_settings = FALSE, exclusions = exclusions)
-    }
+    cli::cli_progress_along(files, name = "Running linters"),
+    function(idx) lint(files[idx], ..., parse_settings = FALSE, exclusions = exclusions)
```
Comment (IndrajeetPatil) [inline R/lint.R:187]: cf. https://cli.r-lib.org/articles/progress.html#progress-bars-for-mapping-functions-cli_progress_along

**Comment by AshesITR** [comment]
Re testing: Is there a way to hook into progress bar updates, or supply a "testing" progress bar to cli?

Otherwise this seems hard to properly and robustly test - especially the animation.

**Comment by MichaelChirico** [comment]
I don't see a need to test thing (formally/in testthat suite), besides ensuring it doesn't cause any error. We can rely on user reports of things "looking bad" IMO.

**Comment by IndrajeetPatil** [comment]
Technically, this is ready for a review.

The only thing I am unsure of is whether we should be using an experimental feature from `{cli}`. Yes, it's experimental for the last 3 years. Still, the changes are so minimal that, in a worst-case—albeit improbable—scenario of `{cli}` deprecating the feature, it'd be trivial to revert this commit and return to using `{utils}`.

**Comment by MichaelChirico** [comment]
@gaborcsardi mind weighing in on the current lifecycle status of `cli_progress_along()`? "it's still experimental" is also fine, just want to be sure we're deciding on fully up-to-date info here.

**Comment by IndrajeetPatil** [comment]
I had already confirmed this with Gabor:

https://github.com/r-lib/cli/issues/709

**Comment by MichaelChirico** [comment]
Thanks, I missed the linked issue!

File: NEWS.md [hunk 2]
```diff
@@ -91,6 +91,7 @@
 
 ## Notes
 
+* All user-facing messages (including progress bar) are now prepared using the `{cli}` package (#2418 and #2641, @IndrajeetPatil). All messages have been reviewed and updated to be more informative and consistent.
```
Comment (MichaelChirico) [inline NEWS.md:94]:
```suggestion
* All user-facing messages (including progress bars) are now prepared using the `{cli}` package (#2418 and #2641, @IndrajeetPatil). All messages have been reviewed and updated to be more informative and consistent.
```

File: R/lint.R [hunk 3]
```diff
@@ -182,20 +182,11 @@ lint_dir <- function(path = ".", ...,
...
-    txtProgressBar(max = length(files), style = 3L)
-  }
-
   lints <- flatten_lints(lapply(
-    files,
-    function(file) {
-      maybe_report_progress(pb)
-      lint(file, ..., parse_settings = FALSE, exclusions = exclusions)
-    }
+    cli::cli_progress_along(files, name = "Running linters"),
```
Comment (MichaelChirico) [inline R/lint.R:186]:
```suggestion
    # NB: This cli API is experimental (https://github.com/r-lib/cli/issues/709)
    cli::cli_progress_along(files, name = "Running linters"),
```

File: R/lint.R [hunk 4]
```diff
@@ -182,20 +182,12 @@ lint_dir <- function(path = ".", ...,
...
-  }
-
   lints <- flatten_lints(lapply(
-    files,
-    function(file) {
-      maybe_report_progress(pb)
-      lint(file, ..., parse_settings = FALSE, exclusions = exclusions)
-    }
+    # NB: This cli API is experimental (https://github.com/r-lib/cli/issues/709)
+    cli::cli_progress_along(files, name = "Running linters"),
```
Comment (MichaelChirico) [inline R/lint.R:187]: is this no longer conditional on `isTRUE(show_progress)`?

Comment (IndrajeetPatil) [reply R/lint.R]: Good catch!

- `show_progress = TRUE`

Screenshot 2024-12-02 at 08 49 13

- `show_progress = FALSE`

Screenshot 2024-12-02 at 08 45 45

**Review by MichaelChirico — APPROVED** [review]

File: R/lint.R [hunk 5]
```diff
@@ -182,20 +182,24 @@ lint_dir <- function(path = ".", ...,
...
+      # NB: This cli API is experimental (https://github.com/r-lib/cli/issues/709)
+      cli::cli_progress_along(files, name = "Running linters"),
+      function(idx) {
+        lint(files[idx], ..., parse_settings = FALSE, exclusions = exclusions)
+      }
+    )
+  } else {
+    lints <- lapply(
+      files,
+      function(file) {
```
Comment (IndrajeetPatil) [inline R/lint.R:196]: Should I just no lint this?

It's catch-22 situation:

- if I use anonymous function, it triggers `unnecessary_lambda_linter()` lint
- if I don't, it leads to `Error:  The use of linters of class 'function' was deprecated in lintr version 3.0.0` error

Comment (IndrajeetPatil) [reply R/lint.R]: I decided to nolint it.

**Review by MichaelChirico — APPROVED** [review]

**Comment by MichaelChirico** [comment]
Oops, just noticed my comment wound up in pending state and only got sent now! Sorry...

**Pull request merged**